apps/emqx_resource/src/emqx_resource_worker.erl

```diff
@@ -499,28 +519,39 @@ do_flush(Data0, #{is_batch := true, batch := Batch, ack_ref := QAckRef, new_queu
         %% failed and is not async; keep the request in the queue to
         %% be retried
         {true, false} ->
-            {next_state, blocked, Data1};
+            %% Note: currently, we cannot safely pop an item from
+            %% `replayq', keep the old reference to the queue and
+            %% later try to append new items to the old ref: by
+            %% popping an item, we may cause the side effect of
+            %% closing an open segment and opening a new one, and the
+            %% later `append' with the old file descriptor will fail
+            %% with `einval' because it has been closed.  So we are
+            %% forced to re-append the item, changing the order of
+            %% requests...
+            ok = replayq:ack(Q1, QAckRef),
+            SentBatch = mark_as_sent(Batch),
+            Q2 = append_queue(Id, Index, Q1, SentBatch),
+            Data2 = Data1#{queue := Q2},
+            {next_state, blocked, Data2};
```

Comment:
is this because we do not have the payload in the inflight table, so we are forced to re-append ?
can we, in this case, update the ets table with payload, so we can retry off from there ?

Reply:
In this clause, the request is not async, which means it's not in the inflight table.  That case is the following `{true, true}` clause.  In that case, we also ack it in replayq and it'll be kept in the ETS table to be retried.

Reply:
Ah, maybe you mean to use the inflight table also for sync requests in this case when we enter the `blocked` state?

That was one of my questions: if the original intention was that only async requests should populate the inflight table.  

I think for that to work, we will need to change the way the requests are enqueued.  Before processing a batch, we'd need to also add the requests to the inflight table, so that we could drop the request from the queue (as we already do for async requests) and also keep the order while retrying.  

That would mean that sync requests would start to take space in the inflight table, and I don't know if that was the original intention of that table usage. :thinking: